File: 654.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* constructMaximumBinaryTree(vector<int>& nums) {
        return constructMaximumBinaryTree(nums, 0, nums.size() - 1);
    }
private:
    TreeNode* constructMaximumBinaryTree(vector<int>& nums, int left, int right){
        if(left > right)
            return NULL;
        if(left == right)
            return new TreeNode(nums[left]);
        int index = left;
        int max = nums[left];
        for(int i = left + 1; i <= right; i++)
            if(nums[i] > max){
                index = i;
                max = nums[i];
            }
        TreeNode* root = new TreeNode(max);
        root->left = constructMaximumBinaryTree(nums, left, index - 1);
        root->right = constructMaximumBinaryTree(nums, index + 1, right);
        return root;
        
    }
};

File: 654.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def newConstructMaximumBinaryTree(self, nums: List[int], left, right) -> TreeNode:
        if left > right:
            return None
        if left == right:
            return TreeNode(nums[left])
        index = left;
        max = nums[left]
        
        for i in range(left + 1, right + 1):
            if nums[i] > max:
                max = nums[i]
                index = i
        
        root = TreeNode(max)
        root.left = self.newConstructMaximumBinaryTree(nums, left, index - 1)
        root.right = self.newConstructMaximumBinaryTree(nums, index + 1, right)
        return root
    
    def constructMaximumBinaryTree(self, nums: List[int]) -> TreeNode:
        return self.newConstructMaximumBinaryTree(nums, 0, len(nums) - 1)
